src/cpp/session/modules/clang/CodeCompletion.cpp

```diff
@@ -154,6 +163,244 @@ core::json::Object toJson(const CodeCompleteResult& result)
    return resultJson;
 }
 
+void discoverTranslationUnitIncludePaths(const FilePath& filePath,
+                                         std::vector<std::string>* pIncludePaths)
+{
+   std::vector<std::string> args =
+         rCompilationDatabase().compileArgsForTranslationUnit(
+            filePath.absolutePathNative(), false);
+   
+   BOOST_FOREACH(const std::string& arg, args)
+   {
+      if (boost::algorithm::starts_with(arg, "-I"))
+      {
+         // skip RStudio's libclang builtin header paths
+         if (arg.find("libclang/builtin-headers") != std::string::npos)
+            continue;
+         
+         pIncludePaths->push_back(
+                  string_utils::strippedOfQuotes(arg.substr(2)));
+      }
+   }
+}
+
+void discoverSystemIncludePaths(std::vector<std::string>* pIncludePaths)
+{
+   core::system::ProcessOptions processOptions;
+   processOptions.redirectStdErrToStdOut = true;
+   
+   // resolve R CMD location for shell command
+   FilePath rBinDir;
+   Error error = module_context::rBinDir(&rBinDir);
+   if (error)
+   {
+      LOG_ERROR(error);
+      return;
+   }
+   
+   shell_utils::ShellCommand rCmd = module_context::rCmd(rBinDir);
+   
+   // get the CXX compiler by asking R
+   std::string compilerPath;
```

Comment:
I took a look at this as well and it was a bit difficult to dis-entangle; now that we're caching the results of our lookup I think we don't need to worry about this.